### ISSUE TITLE
LANDER/TRE Intermittent Launch Failure 

### ISSUE THREAD

**Issue by timhowcroft** [lsc-sde/lsc-sde#93]
![image](https://github.com/user-attachments/assets/b8d12f05-7052-40a4-b5da-85f0a2863e66)
Frequently when trying to launched LANDER, especially for the first time in a few days, the request times out. Often resolves itself once Vishnu has managed to log in.

Similar failures reported by UoM students.

**Comment by qcaas-nhs-sjt** [comment]
hi @timhowcroft  I've been asked to investigate this issue, could you confirm which workspace you are using?

**Comment by timhowcroft** [comment]
Hi @qcaas-nhs-sjt it's usually `lth-data-science-team-workspace` that has the issues, but I believe I may have had it one a couple of the others.

**Comment by qcaas-nhs-sjt** [comment]
Log Extract:

```
[E 2024-10-08 10:10:30.227 JupyterHub spawner:2810] Pod jupyterhub/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk never showed up in reflector, restarting pod reflector
[E 2024-10-08 10:10:30.227 JupyterHub spawner:2814] Pods: []
[W 2024-10-08 10:10:30.228 JupyterHub user:876] timothy.howcroft@lthtr.nhs.uk's server failed to start in 900 seconds, giving up.
    
    Common causes of this timeout, and debugging tips:
    
    1. Everything is working, but it took too long.
       To fix: increase `Spawner.start_timeout` configuration
       to a number of seconds that is enough for spawners to finish starting.
    2. The server didn't finish starting,
       or it crashed due to a configuration issue.
       Check the single-user server's logs for hints at what needs fixing.
    
[W 2024-10-08 10:10:30.271 JupyterHub spawner:2250] Pod jupyterhub/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk url changed! jupyterhub-hub-5759c44697-tcpbg:40663 -> 100.64.0.112:8888
[W 2024-10-08 10:10:30.272 JupyterHub spawner:170] 
    The shared database session at Spawner.db is deprecated, and will be removed.
    Please manage your own database and connections.
    
    Contact JupyterHub at https://github.com/jupyterhub/jupyterhub/issues/3700
    if you have questions or ideas about direct database needs for your Spawner.
    
[I 2024-10-08 10:10:30.287 JupyterHub reflector:282] watching for pods with label selector='component=singleuser-server' in namespace jupyterhub
[I 2024-10-08 10:10:30.289 JupyterHub spawner:2849] Deleting pod jupyterhub/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk
[E 2024-10-08 10:10:32.877 JupyterHub gen:630] Exception in Future <Task finished name='Task-288394' coro=<BaseHandler.spawn_single_user.<locals>.finish_user_spawn() done, defined at /usr/local/lib/python3.11/site-packages/jupyterhub/handlers/base.py:981> exception=TimeoutError('pod jupyterhub/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk did not start in 900 seconds!')> after timeout
    Traceback (most recent call last):
      File "/usr/local/lib/python3.11/site-packages/tornado/gen.py", line 625, in error_callback
        future.result()
      File "/usr/local/lib/python3.11/site-packages/jupyterhub/handlers/base.py", line 988, in finish_user_spawn
        await spawn_future
      File "/usr/local/lib/python3.11/site-packages/jupyterhub/user.py", line 902, in spawn
        raise e
      File "/usr/local/lib/python3.11/site-packages/jupyterhub/user.py", line 798, in spawn
        url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/kubespawner/spawner.py", line 2801, in _start
        await exponential_backoff(
      File "/usr/local/lib/python3.11/site-packages/jupyterhub/utils.py", line 237, in exponential_backoff
        raise asyncio.TimeoutError(fail_message)
    TimeoutError: pod jupyterhub/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk did not start in 900 seconds!
 ```

**Comment by qcaas-nhs-sjt** [comment]
events pertaining to the pod:
```
46m         Normal   Scheduled          pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Successfully assigned jupyterhub/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk to aks-default-33462530-vmss000000
46m         Normal   Pulled             pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Container image "quay.io/jupyterhub/k8s-network-tools:3.2.1" already present on machine
46m         Normal   Created            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Created container block-cloud-metadata
46m         Normal   Started            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Started container block-cloud-metadata
46m         Normal   Pulled             pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Container image "lscsde/datascience-notebook-default:0.1.0" already present on machine
46m         Normal   Created            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Created container notebook
46m         Normal   Started            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Started container notebook
32m         Normal   Killing            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Stopping container notebook
28m         Normal   Scheduled          pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Successfully assigned jupyterhub/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk to aks-default-33462530-vmss000000
28m         Normal   Pulled             pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Container image "quay.io/jupyterhub/k8s-network-tools:3.2.1" already present on machine
28m         Normal   Created            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Created container block-cloud-metadata
28m         Normal   Started            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Started container block-cloud-metadata
28m         Normal   Pulled             pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Container image "lscsde/datascience-notebook-default:0.1.0" already present on machine
28m         Normal   Created            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Created container notebook
28m         Normal   Started            pod/jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk      Started container notebook
```

Seems that according to k8s, the pod started successfully but then was killed 14-15 minutes later which would be in line with the timeout of 900 seconds, according to the jupyter logs in the previous update because it didn't report in to jupyter hub in the 900 seconds it was deleted by jupyter hub.

My instincts tell me that this is a problem with jupyterhub / jupyter notebook rather than anything with us but going to be difficult to confirm this I've tried looking at the previous container logs for the user pod but they aren't available from the control plane, will see if these have made it into the azure diagnostics workspace.

We could try increasing the timeout as per the error message, though 900 seconds seems a very long time for the pod to come up unless there was an autoscaler event. I will look into this possibility as well though looking at the current usage stats i doubt that this would be the case.

There is an issue which some users have reported where this can happen:
https://github.com/jupyterhub/kubespawner/issues/577

will keep investigating

**Comment by qcaas-nhs-sjt** [comment]
Container logs:
```
LogEntry
"[I 2024-10-08 09:56:34.211 ServerApp] Package jupyterhub took 0.0000s to import"
"[I 2024-10-08 09:56:34.223 ServerApp] Package jupyter_lsp took 0.0124s to import"
"[W 2024-10-08 09:56:34.224 ServerApp] A `_jupyter_server_extension_points` function was not found in jupyter_lsp. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server."
"[I 2024-10-08 09:56:34.225 ServerApp] Package jupyter_server_mathjax took 0.0010s to import"
"[I 2024-10-08 09:56:34.293 ServerApp] Package jupyter_server_proxy took 0.0675s to import"
"[I 2024-10-08 09:56:34.298 ServerApp] Package jupyter_server_terminals took 0.0046s to import"
"[I 2024-10-08 09:56:34.371 ServerApp] Package jupyterlab took 0.0725s to import"
"[I 2024-10-08 09:56:35.700 ServerApp] Package jupyterlab_code_formatter took 0.6172s to import"
"[I 2024-10-08 09:56:35.719 ServerApp] Package jupyterlab_git took 0.0190s to import"
"[I 2024-10-08 09:56:35.722 ServerApp] Package nbclassic took 0.0024s to import"
"[W 2024-10-08 09:56:35.724 ServerApp] A `_jupyter_server_extension_points` function was not found in nbclassic. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server."
"[I 2024-10-08 09:56:35.724 ServerApp] Package nbdime took 0.0000s to import"
"[I 2024-10-08 09:56:35.724 ServerApp] Package notebook took 0.0000s to import"
"[I 2024-10-08 09:56:35.726 ServerApp] Package notebook_shim took 0.0000s to import"
"[W 2024-10-08 09:56:35.726 ServerApp] A `_jupyter_server_extension_points` function was not found in notebook_shim. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server."
"[I 2024-10-08 09:56:36.577 ServerApp] Package panel.io.jupyter_server_extension took 0.8504s to import"
"[I 2024-10-08 09:56:36.578 ServerApp] jupyter_lsp | extension was successfully linked."
"[I 2024-10-08 09:56:36.582 ServerApp] jupyter_server_mathjax | extension was successfully linked."
"[I 2024-10-08 09:56:36.582 ServerApp] jupyter_server_proxy | extension was successfully linked."
"[I 2024-10-08 09:56:36.585 ServerApp] jupyter_server_terminals | extension was successfully linked."
"[I 2024-10-08 09:56:36.586 JupyterHubSingleUser] Starting jupyterhub single-user server extension version 4.0.2"
"[I 2024-10-08 09:56:36.586 JupyterHubSingleUser] Using default url from environment $JUPYTERHUB_DEFAULT_URL: /lab"
"[I 2024-10-08 09:56:36.589 ServerApp] jupyterhub | extension was successfully linked."
"[W 2024-10-08 09:56:36.590 LabApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose."
"[I 2024-10-08 09:56:36.593 ServerApp] jupyterlab | extension was successfully linked."
"[I 2024-10-08 09:56:36.593 ServerApp] jupyterlab_code_formatter | extension was successfully linked."
"[I 2024-10-08 09:56:36.593 ServerApp] jupyterlab_git | extension was successfully linked."
"[W 2024-10-08 09:56:36.595 NotebookApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose."
"[I 2024-10-08 09:56:36.597 ServerApp] nbclassic | extension was successfully linked."
"[I 2024-10-08 09:56:36.597 ServerApp] nbdime | extension was successfully linked."
"[W 2024-10-08 09:56:36.598 JupyterNotebookApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose."
"[I 2024-10-08 09:56:36.601 ServerApp] notebook | extension was successfully linked."
"[I 2024-10-08 09:56:36.601 ServerApp] Writing Jupyter server cookie secret to /tmp/jupyter_cookie_secret"
"[I 2024-10-08 09:56:36.903 ServerApp] notebook_shim | extension was successfully linked."
"[I 2024-10-08 09:56:36.903 ServerApp] panel.io.jupyter_server_extension | extension was successfully linked."
"[I 2024-10-08 09:56:36.920 ServerApp] notebook_shim | extension was successfully loaded."
"[I 2024-10-08 09:56:36.921 ServerApp] jupyter_lsp | extension was successfully loaded."
"[I 2024-10-08 09:56:36.922 ServerApp] jupyter_server_mathjax | extension was successfully loaded."
"[I 2024-10-08 09:56:36.937 ServerApp] jupyter_server_proxy | extension was successfully loaded."
"[I 2024-10-08 09:56:36.938 ServerApp] jupyter_server_terminals | extension was successfully loaded."
"[I 2024-10-08 09:56:36.946 JupyterHubSingleUser] Updating Hub with activity every 300 seconds"
"[I 2024-10-08 09:56:36.946 ServerApp] jupyterhub | extension was successfully loaded."
"[I 2024-10-08 09:56:36.966 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.11/site-packages/jupyterlab"
"[I 2024-10-08 09:56:36.966 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab"
"[I 2024-10-08 09:56:36.966 LabApp] Extension Manager is 'pypi'."
"[I 2024-10-08 09:56:36.973 ServerApp] jupyterlab | extension was successfully loaded."
"[I 2024-10-08 09:56:36.974 ServerApp] Registered jupyterlab_code_formatter server extension"
"[I 2024-10-08 09:56:36.974 ServerApp] jupyterlab_code_formatter | extension was successfully loaded."
"[I 2024-10-08 09:56:36.979 ServerApp] jupyterlab_git | extension was successfully loaded."
"[I 2024-10-08 09:56:36.982 ServerApp] nbclassic | extension was successfully loaded."
"[I 2024-10-08 09:56:37.080 ServerApp] nbdime | extension was successfully loaded."
"[I 2024-10-08 09:56:37.084 ServerApp] notebook | extension was successfully loaded."
"[I 2024-10-08 09:56:37.085 ServerApp] panel.io.jupyter_server_extension | extension was successfully loaded."
"[I 2024-10-08 09:56:37.085 ServerApp] Serving notebooks from local directory: /home/jovyan"
"[I 2024-10-08 09:56:37.085 ServerApp] Jupyter Server 2.7.3 is running at:"
"[I 2024-10-08 09:56:37.085 ServerApp] http://jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk:8888/user/[redacted]/lab?token=..."
"[I 2024-10-08 09:56:37.085 ServerApp]     http://127.0.0.1:8888/user/[redacted]/lab?token=..."
"[I 2024-10-08 09:56:37.085 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation)."
"[I 2024-10-08 09:56:38.227 ServerApp] Skipped non-installed server(s): bash-language-server, dockerfile-language-server-nodejs, javascript-typescript-langserver, jedi-language-server, julia-language-server, pyright, python-language-server, python-lsp-server, r-languageserver, sql-language-server, texlab, typescript-language-server, unified-language-server, vscode-css-languageserver-bin, vscode-html-languageserver-bin, vscode-json-languageserver-bin, yaml-language-server"
"[C 2024-10-08 10:10:30.357 ServerApp] received signal 15, stopping"
"[I 2024-10-08 10:10:30.357 ServerApp] Shutting down 13 extensions"

"  _   _          _      _"
" | | | |_ __  __| |__ _| |_ ___"
" | |_| | '_ \/ _` / _` |  _/ -_)"
"  \___/| .__/\__,_\__,_|\__\___|"
"       |_|"
"                                                                           "
"Read the migration plan to Notebook 7 to learn about the new features and the actions to take if you are using extensions."

"https://jupyter-notebook.readthedocs.io/en/latest/migrate_to_notebook7.html"

"Please note that updating to Notebook 7 might break some of your extensions."

"[I 2024-10-08 10:14:04.433 ServerApp] Package jupyterhub took 0.0000s to import"
"[I 2024-10-08 10:14:04.443 ServerApp] Package jupyter_lsp took 0.0096s to import"
"[W 2024-10-08 10:14:04.443 ServerApp] A `_jupyter_server_extension_points` function was not found in jupyter_lsp. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server."
"[I 2024-10-08 10:14:04.444 ServerApp] Package jupyter_server_mathjax took 0.0009s to import"
"[I 2024-10-08 10:14:04.506 ServerApp] Package jupyter_server_proxy took 0.0609s to import"
"[I 2024-10-08 10:14:04.511 ServerApp] Package jupyter_server_terminals took 0.0046s to import"
"[I 2024-10-08 10:14:04.566 ServerApp] Package jupyterlab took 0.0552s to import"
"[I 2024-10-08 10:14:05.736 ServerApp] Package jupyterlab_code_formatter took 0.5549s to import"
"[I 2024-10-08 10:14:05.751 ServerApp] Package jupyterlab_git took 0.0146s to import"
"[I 2024-10-08 10:14:05.754 ServerApp] Package nbclassic took 0.0020s to import"
"[W 2024-10-08 10:14:05.755 ServerApp] A `_jupyter_server_extension_points` function was not found in nbclassic. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server."
"[I 2024-10-08 10:14:05.756 ServerApp] Package nbdime took 0.0000s to import"
"[I 2024-10-08 10:14:05.756 ServerApp] Package notebook took 0.0000s to import"
"[I 2024-10-08 10:14:05.758 ServerApp] Package notebook_shim took 0.0000s to import"
"[W 2024-10-08 10:14:05.758 ServerApp] A `_jupyter_server_extension_points` function was not found in notebook_shim. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server."
"[I 2024-10-08 10:14:06.560 ServerApp] Package panel.io.jupyter_server_extension took 0.8015s to import"
"[I 2024-10-08 10:14:06.560 ServerApp] jupyter_lsp | extension was successfully linked."
"[I 2024-10-08 10:14:06.564 ServerApp] jupyter_server_mathjax | extension was successfully linked."
"[I 2024-10-08 10:14:06.564 ServerApp] jupyter_server_proxy | extension was successfully linked."
"[I 2024-10-08 10:14:06.567 ServerApp] jupyter_server_terminals | extension was successfully linked."
"[I 2024-10-08 10:14:06.568 JupyterHubSingleUser] Starting jupyterhub single-user server extension version 4.0.2"
"[I 2024-10-08 10:14:06.568 JupyterHubSingleUser] Using default url from environment $JUPYTERHUB_DEFAULT_URL: /lab"
"[I 2024-10-08 10:14:06.571 ServerApp] jupyterhub | extension was successfully linked."
"[W 2024-10-08 10:14:06.572 LabApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose."
"[I 2024-10-08 10:14:06.575 ServerApp] jupyterlab | extension was successfully linked."
"[I 2024-10-08 10:14:06.575 ServerApp] jupyterlab_code_formatter | extension was successfully linked."
"[I 2024-10-08 10:14:06.575 ServerApp] jupyterlab_git | extension was successfully linked."
"[W 2024-10-08 10:14:06.577 NotebookApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose."
"[I 2024-10-08 10:14:06.579 ServerApp] nbclassic | extension was successfully linked."
"[I 2024-10-08 10:14:06.579 ServerApp] nbdime | extension was successfully linked."
"[W 2024-10-08 10:14:06.580 JupyterNotebookApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose."
"[I 2024-10-08 10:14:06.583 ServerApp] notebook | extension was successfully linked."
"[I 2024-10-08 10:14:06.583 ServerApp] Writing Jupyter server cookie secret to /tmp/jupyter_cookie_secret"
"[I 2024-10-08 10:14:06.856 ServerApp] notebook_shim | extension was successfully linked."
"[I 2024-10-08 10:14:06.857 ServerApp] panel.io.jupyter_server_extension | extension was successfully linked."
"[I 2024-10-08 10:14:06.873 ServerApp] notebook_shim | extension was successfully loaded."
"[I 2024-10-08 10:14:06.875 ServerApp] jupyter_lsp | extension was successfully loaded."
"[I 2024-10-08 10:14:06.875 ServerApp] jupyter_server_mathjax | extension was successfully loaded."
"[I 2024-10-08 10:14:06.891 ServerApp] jupyter_server_proxy | extension was successfully loaded."
"[I 2024-10-08 10:14:06.891 ServerApp] jupyter_server_terminals | extension was successfully loaded."
"[I 2024-10-08 10:14:06.899 JupyterHubSingleUser] Updating Hub with activity every 300 seconds"
"[I 2024-10-08 10:14:06.899 ServerApp] jupyterhub | extension was successfully loaded."
"[I 2024-10-08 10:14:06.913 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.11/site-packages/jupyterlab"
"[I 2024-10-08 10:14:06.914 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab"
"[I 2024-10-08 10:14:06.914 LabApp] Extension Manager is 'pypi'."
"[I 2024-10-08 10:14:06.917 ServerApp] jupyterlab | extension was successfully loaded."
"[I 2024-10-08 10:14:06.917 ServerApp] Registered jupyterlab_code_formatter server extension"
"[I 2024-10-08 10:14:06.917 ServerApp] jupyterlab_code_formatter | extension was successfully loaded."
"[I 2024-10-08 10:14:06.922 ServerApp] jupyterlab_git | extension was successfully loaded."
"[I 2024-10-08 10:14:06.930 ServerApp] nbclassic | extension was successfully loaded."
"[I 2024-10-08 10:14:07.003 ServerApp] nbdime | extension was successfully loaded."
"[I 2024-10-08 10:14:07.007 ServerApp] notebook | extension was successfully loaded."
"[I 2024-10-08 10:14:07.008 ServerApp] panel.io.jupyter_server_extension | extension was successfully loaded."
"[I 2024-10-08 10:14:07.008 ServerApp] Serving notebooks from local directory: /home/jovyan"
"[I 2024-10-08 10:14:07.008 ServerApp] Jupyter Server 2.7.3 is running at:"
"[I 2024-10-08 10:14:07.008 ServerApp] http://jupyter-timothy-2ehowcroft-40lthtr-2enhs-2euk:8888/user/[redacted]/lab?token=..."
"[I 2024-10-08 10:14:07.008 ServerApp]     http://127.0.0.1:8888/user/[redacted]/lab?token=..."
"[I 2024-10-08 10:14:07.008 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation)."
"[I 2024-10-08 10:14:07.110 ServerApp] 302 GET /user/[redacted]/ -> /user/[redacted]/lab? (@100.64.0.108) 0.68ms"
"[I 2024-10-08 10:14:07.209 ServerApp] 302 GET /user/[redacted]/ -> /user/[redacted]/lab? (@10.26.104.12) 0.70ms"
"[I 2024-10-08 10:14:07.273 ServerApp] 302 GET /user/[redacted]/lab? -> /hub/api/oauth2/authorize?client_id=jupyterhub-user-timothy.howcroft%2540lthtr.nhs.uk&redirect_uri=%2Fuser%2Ftimothy.howcroft%40lthtr.nhs.uk%2Foauth_callback&response_type=code&state=[secret] (@10.26.104.12) 22.29ms"
"[I 2024-10-08 10:14:07.486 ServerApp] Logged-in user {'admin': False, 'groups': [], 'kind': 'user', 'name': '[redacted]', 'session_id': 'edeff0fa5f634565b15199017f0b8bd6', 'scopes': ['access:servers!server=[redacted]/', 'read:users:groups!user=[redacted]', 'read:users:name!user=[redacted]']}"
"[I 2024-10-08 10:14:07.488 ServerApp] 302 GET /user/[redacted]/oauth_callback?code=[secret]&state=[secret] -> /user/[redacted]/lab? (@10.26.104.12) 104.64ms"
"[I 2024-10-08 10:14:07.580 ServerApp] 200 GET /user/[redacted]/lab? ([redacted]@10.26.104.12) 49.57ms"
"[I 2024-10-08 10:14:08.567 ServerApp] 200 GET /user/[redacted]/api/me?1728382448461 ([redacted]@10.26.104.12) 1.08ms"
"[I 2024-10-08 10:14:08.219 ServerApp] Skipped non-installed server(s): bash-language-server, dockerfile-language-server-nodejs, javascript-typescript-langserver, jedi-language-server, julia-language-server, pyright, python-language-server, python-lsp-server, r-languageserver, sql-language-server, texlab, typescript-language-server, unified-language-server, vscode-css-languageserver-bin, vscode-html-languageserver-bin, vscode-json-languageserver-bin, yaml-language-server"
"0.00s - Debugger warning: It seems that frozen modules are being used, which may"
"0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off"
"0.00s - to python to disable frozen modules."
"0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation."
"[I 2024-10-08 10:14:08.565 ServerApp] 200 GET /user/[redacted]/api/kernelspecs?1728382448461 ([redacted]@10.26.104.12) 97.63ms"
"[I 2024-10-08 10:14:08.640 ServerApp] 200 GET /user/[redacted]/lab/api/settings?1728382448465 ([redacted]@10.26.104.12) 72.35ms"
"[I 2024-10-08 10:14:08.641 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382448469 ([redacted]@10.26.104.12) 72.01ms"
"[I 2024-10-08 10:14:08.642 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382448469 ([redacted]@10.26.104.12) 72.29ms"
"[I 2024-10-08 10:14:08.643 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382448469 ([redacted]@10.26.104.12) 73.20ms"
"[I 2024-10-08 10:14:08.645 ServerApp] 200 GET /user/[redacted]/api/kernelspecs?1728382448599 ([redacted]@10.26.104.12) 1.98ms"
"[I 2024-10-08 10:14:08.647 ServerApp] 101 GET /user/[redacted]/api/events/subscribe?token=[secret] ([redacted]@10.26.104.11) 0.84ms"
"[I 2024-10-08 10:14:08.648 ServerApp] 200 GET /user/[redacted]/api/me?1728382448601 ([redacted]@10.26.104.12) 0.75ms"
"[I 2024-10-08 10:14:09.863 ServerApp] 200 GET /user/[redacted]/lab/api/translations/default?1728382449829 ([redacted]@10.26.104.12) 25.52ms"
"[I 2024-10-08 10:14:09.918 ServerApp] 200 GET /user/[redacted]/api/config/jupyterlabapputilsextensionannouncements?1728382449906 ([redacted]@10.26.104.12) 2.70ms"
"[I 2024-10-08 10:14:09.973 ServerApp] 200 GET /user/[redacted]/lsp/status?1728382449963 ([redacted]@10.26.104.12) 2.74ms"
"[I 2024-10-08 10:14:09.991 ServerApp] 200 GET /user/[redacted]/server-proxy/servers-info?1728382449983 ([redacted]@10.26.104.12) 2.31ms"
"[I 2024-10-08 10:14:10.003 ServerApp] 200 GET /user/[redacted]/git/settings?version=0.50.0&1728382449980 ([redacted]@10.26.104.12) 16.11ms"
"[I 2024-10-08 10:14:10.021 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382449978 ([redacted]@10.26.104.12) 28.40ms"
"[I 2024-10-08 10:14:10.077 ServerApp] 200 GET /user/[redacted]/lab/api/settings/@jupyterlab/codemirror-extension:plugin?1728382450016 ([redacted]@10.26.104.12) 3.27ms"
"[I 2024-10-08 10:14:10.084 ServerApp] 200 GET /user/[redacted]/lab/api/settings/@jupyterlab/notebook-extension:panel?1728382450019 ([redacted]@10.26.104.12) 6.92ms"
"[I 2024-10-08 10:14:10.092 ServerApp] 200 GET /user/[redacted]/lab/api/settings/jupyterlab_code_formatter:settings?1728382450020 ([redacted]@10.26.104.12) 13.84ms"
"[I 2024-10-08 10:14:10.094 ServerApp] 200 GET /user/[redacted]/lab/api/workspaces/default?1728382450069 ([redacted]@10.26.104.12) 0.85ms"
"[I 2024-10-08 10:14:10.103 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382450087 ([redacted]@10.26.104.12) 7.87ms"
"[I 2024-10-08 10:14:10.713 ServerApp] 200 GET /user/[redacted]/jupyterlab_code_formatter/formatters?1728382450152 ([redacted]@10.26.104.12) 540.19ms"
"[I 2024-10-08 10:14:10.715 ServerApp] 200 GET /user/[redacted]/lab/api/translations?1728382449974 ([redacted]@10.26.104.12) 733.36ms"
"[I 2024-10-08 10:14:10.776 ServerApp] 200 GET /user/[redacted]/lab/api/settings?ids_only=true&1728382450167 ([redacted]@10.26.104.12) 60.83ms"
"[I 2024-10-08 10:14:10.778 ServerApp] 204 PUT /user/[redacted]/lab/api/workspaces/default?1728382450705 ([redacted]@10.26.104.12) 1.40ms"
"[I 2024-10-08 10:14:10.825 ServerApp] 200 POST /user/[redacted]/git/show_prefix?1728382450813 ([redacted]@10.26.104.12) 5.20ms"
"[I 2024-10-08 10:14:10.831 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382450814 ([redacted]@10.26.104.12) 9.41ms"
"[I 2024-10-08 10:14:10.884 ServerApp] 200 GET /user/[redacted]/server-proxy/icon/vscode ([redacted]@10.26.104.12) 1.09ms"
"[I 2024-10-08 10:14:10.886 ServerApp] 200 GET /user/[redacted]/server-proxy/icon/rstudio ([redacted]@10.26.104.12) 1.36ms"
"[I 2024-10-08 10:14:11.023 LabApp] Build is up to date"
"[I 2024-10-08 10:14:11.024 ServerApp] 200 GET /user/[redacted]/lab/api/build?1728382449927 ([redacted]@10.26.104.12) 1085.09ms"
"[I 2024-10-08 10:14:11.395 ServerApp] 204 PUT /user/[redacted]/lab/api/workspaces/default?1728382451386 ([redacted]@10.26.104.12) 1.30ms"
"[I 2024-10-08 10:14:18.701 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382458689 ([redacted]@10.26.104.12) 1.23ms"
"[I 2024-10-08 10:14:18.729 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382458720 ([redacted]@10.26.104.12) 0.90ms"
"[I 2024-10-08 10:14:18.759 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382458751 ([redacted]@10.26.104.12) 1.03ms"
"[I 2024-10-08 10:14:20.890 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382460876 ([redacted]@10.26.104.12) 6.28ms"
"[I 2024-10-08 10:14:28.757 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382468738 ([redacted]@10.26.104.12) 1.02ms"
"[I 2024-10-08 10:14:28.777 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382468770 ([redacted]@10.26.104.12) 0.86ms"
"[I 2024-10-08 10:14:28.803 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382468794 ([redacted]@10.26.104.12) 1.24ms"
"[I 2024-10-08 10:14:30.945 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382470930 ([redacted]@10.26.104.12) 6.71ms"
"[I 2024-10-08 10:14:38.808 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382478795 ([redacted]@10.26.104.12) 1.07ms"
"[I 2024-10-08 10:14:38.833 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382478826 ([redacted]@10.26.104.12) 0.87ms"
"[I 2024-10-08 10:14:38.848 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382478842 ([redacted]@10.26.104.12) 0.94ms"
"[I 2024-10-08 10:14:41.022 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382480996 ([redacted]@10.26.104.12) 18.81ms"
"[I 2024-10-08 10:14:48.870 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382488851 ([redacted]@10.26.104.12) 1.05ms"
"[I 2024-10-08 10:14:48.890 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382488881 ([redacted]@10.26.104.12) 0.78ms"
"[I 2024-10-08 10:14:48.903 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382488896 ([redacted]@10.26.104.12) 0.87ms"
"[I 2024-10-08 10:14:51.073 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382491058 ([redacted]@10.26.104.12) 6.59ms"
"[I 2024-10-08 10:14:55.337 ServerApp] 204 PUT /user/[redacted]/lab/api/workspaces/default?1728382495328 ([redacted]@10.26.104.12) 1.42ms"
"[I 2024-10-08 10:14:59.324 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382499315 ([redacted]@10.26.104.12) 1.10ms"
"[I 2024-10-08 10:14:59.326 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382499316 ([redacted]@10.26.104.12) 0.75ms"
"[I 2024-10-08 10:14:59.327 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382499316 ([redacted]@10.26.104.12) 0.87ms"
"[I 2024-10-08 10:15:01.341 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382501327 ([redacted]@10.26.104.12) 6.23ms"
"[I 2024-10-08 10:15:10.343 ServerApp] 200 GET /user/[redacted]/api/me?1728382510331 ([redacted]@10.26.104.12) 1.05ms"
"[I 2024-10-08 10:15:10.345 ServerApp] 200 GET /user/[redacted]/api/kernelspecs?1728382510330 ([redacted]@10.26.104.12) 1.81ms"
"[I 2024-10-08 10:16:02.413 ServerApp] 101 GET /user/[redacted]/api/events/subscribe?token=[secret] ([redacted]@10.26.104.13) 1.09ms"
"[I 2024-10-08 10:22:07.537 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382927369 ([redacted]@10.26.104.12) 21.05ms"
"[I 2024-10-08 10:22:07.541 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382927370 ([redacted]@10.26.104.12) 24.25ms"
"[I 2024-10-08 10:22:07.542 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382927370 ([redacted]@10.26.104.12) 24.31ms"
"[I 2024-10-08 10:22:07.556 ServerApp] 200 GET /user/[redacted]/api/contents?content=1&1728382927370 ([redacted]@10.26.104.12) 38.41ms"
"[I 2024-10-08 10:22:07.597 ServerApp] 101 GET /user/[redacted]/api/events/subscribe?token=[secret] ([redacted]@10.26.104.12) 1.16ms"
"[I 2024-10-08 10:22:11.534 ServerApp] 200 GET /user/[redacted]/api/contents/lth-data-science-team-workspace?content=1&1728382931509 ([redacted]@10.26.104.12) 16.18ms"
"[I 2024-10-08 10:22:11.604 ServerApp] 200 POST /user/[redacted]/git/lth-data-science-team-workspace/show_prefix?1728382931569 ([redacted]@10.26.104.12) 29.20ms"
"[I 2024-10-08 10:22:12.084 ServerApp] 204 PUT /user/[redacted]/lab/api/workspaces/default?1728382932076 ([redacted]@10.26.104.12) 1.33ms"
"[I 2024-10-08 10:22:17.588 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382937580 ([redacted]@10.26.104.12) 1.10ms"
"[I 2024-10-08 10:22:17.590 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382937581 ([redacted]@10.26.104.12) 1.13ms"
"[I 2024-10-08 10:22:17.591 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382937581 ([redacted]@10.26.104.12) 1.47ms"
"[I 2024-10-08 10:22:17.618 ServerApp] 200 GET /user/[redacted]/api/contents/lth-data-science-team-workspace?content=1&1728382937596 ([redacted]@10.26.104.12) 16.31ms"
"[I 2024-10-08 10:22:18.343 ServerApp] 200 GET /user/[redacted]/api/me?1728382938336 ([redacted]@10.26.104.12) 1.08ms"
"[I 2024-10-08 10:22:18.345 ServerApp] 200 GET /user/[redacted]/api/kernelspecs?1728382938336 ([redacted]@10.26.104.12) 1.94ms"
"[I 2024-10-08 10:22:28.331 ServerApp] 200 GET /user/[redacted]/api/kernels?1728382948322 ([redacted]@10.26.104.12) 1.89ms"
"[I 2024-10-08 10:22:28.332 ServerApp] 200 GET /user/[redacted]/api/sessions?1728382948323 ([redacted]@10.26.104.12) 1.89ms"
"[I 2024-10-08 10:22:28.333 ServerApp] 200 GET /user/[redacted]/api/terminals?1728382948323 ([redacted]@10.26.104.12) 0.89ms"
"[I 2024-10-08 10:22:28.350 ServerApp] 200 GET /user/[redacted]/api/contents/lth-data-science-team-workspace?content=1&1728382948323 ([redacted]@10.26.104.12) 16.90ms"
"[I 2024-10-08 10:23:20.338 ServerApp] 200 GET /user/[redacted]/api/me?1728383000328 ([redacted]@10.26.104.12) 1.06ms"
"[I 2024-10-08 10:23:20.341 ServerApp] 200 GET /user/[redacted]/api/kernelspecs?1728383000328 ([redacted]@10.26.104.12) 1.93ms"
"[I 2024-10-08 10:23:30.420 ServerApp] 101 GET /user/[redacted]/api/events/subscribe?token=[secret] ([redacted]@10.26.104.11) 1.12ms"

```
looking at these, there is no obvious errors and the container looks like it started the exact same way as it did previously, but for some reason jupyter hub was not aware of the start

**Comment by qcaas-nhs-sjt** [comment]
@manics 

Any pointers you can give me on this? To me it feels like a jupyterhub problem rather than anything to do with networking or the cluster itself which otherwise appears to be functioning normally but you know jupyter/kubespawner better than myself

This environment is using our jupyterhub image lscsde/jupyterhub:0.1.24 which is based upon:

quay.io/jupyterhub/k8s-hub:3.2.1

**Comment by manics** [comment]
I can't see anything obvious- as you've pointed out the singleuser server started but the hub couldn't contact it so killed it after 15 minutes, but on the second attempt it worked fine.

One thing worth trying is to exec into the hub pod, and if it doesn't detect the singleuser server try to wget/curl `http://<pod>:8888` to confirm networking is definitely fine.

@consideRatio do you have any other suggestions?

**Comment by qcaas-nhs-sjt** [comment]
> I can't see anything obvious- as you've pointed out the singleuser server started but the hub couldn't contact it so killed it after 15 minutes, but on the second attempt it worked fine.
Thanks for confirming. I get the feeling that it's going to be one of *those* difficult issues to debug because it's all about timing. I won't get notified about them until the exception is raised at which point jupyter will delete the pod 😆 
 
> One thing worth trying is to exec into the hub pod, and if it doesn't detect the singleuser server try to curl `http://<pod>:8888` to confirm networking is definitely fine.

I'll see if i just keep creating and deleting my own pods, maybe i'll get lucky and reproduce the error myself

**Comment by qcaas-nhs-sjt** [comment]
logs from my pod:
```
$ kubectl logs pods/jupyter-shaun-2eturner1-40nhs-2enet -n jupyterhub 
Defaulted container "notebook" out of: notebook, block-cloud-metadata (init)
[I 2024-10-08 13:13:40.166 ServerApp] Package jupyterhub took 0.0000s to import
[I 2024-10-08 13:13:40.176 ServerApp] Package jupyter_lsp took 0.0097s to import
[W 2024-10-08 13:13:40.176 ServerApp] A `_jupyter_server_extension_points` function was not found in jupyter_lsp. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server.
[I 2024-10-08 13:13:40.177 ServerApp] Package jupyter_server_mathjax took 0.0010s to import
[I 2024-10-08 13:13:40.234 ServerApp] Package jupyter_server_proxy took 0.0570s to import
[I 2024-10-08 13:13:40.239 ServerApp] Package jupyter_server_terminals took 0.0046s to import
[I 2024-10-08 13:13:40.295 ServerApp] Package jupyterlab took 0.0552s to import
[I 2024-10-08 13:13:41.468 ServerApp] Package jupyterlab_code_formatter took 0.5698s to import
[I 2024-10-08 13:13:41.483 ServerApp] Package jupyterlab_git took 0.0147s to import
[I 2024-10-08 13:13:41.486 ServerApp] Package nbclassic took 0.0020s to import
[W 2024-10-08 13:13:41.487 ServerApp] A `_jupyter_server_extension_points` function was not found in nbclassic. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server.
[I 2024-10-08 13:13:41.488 ServerApp] Package nbdime took 0.0000s to import
[I 2024-10-08 13:13:41.488 ServerApp] Package notebook took 0.0000s to import
[I 2024-10-08 13:13:41.489 ServerApp] Package notebook_shim took 0.0000s to import
[W 2024-10-08 13:13:41.490 ServerApp] A `_jupyter_server_extension_points` function was not found in notebook_shim. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server.
[I 2024-10-08 13:13:42.298 ServerApp] Package panel.io.jupyter_server_extension took 0.8084s to import
[I 2024-10-08 13:13:42.299 ServerApp] jupyter_lsp | extension was successfully linked.
[I 2024-10-08 13:13:42.303 ServerApp] jupyter_server_mathjax | extension was successfully linked.
[I 2024-10-08 13:13:42.303 ServerApp] jupyter_server_proxy | extension was successfully linked.
[I 2024-10-08 13:13:42.306 ServerApp] jupyter_server_terminals | extension was successfully linked.
[I 2024-10-08 13:13:42.307 JupyterHubSingleUser] Starting jupyterhub single-user server extension version 4.0.2
[I 2024-10-08 13:13:42.307 JupyterHubSingleUser] Using default url from environment $JUPYTERHUB_DEFAULT_URL: /lab
[I 2024-10-08 13:13:42.310 ServerApp] jupyterhub | extension was successfully linked.
[W 2024-10-08 13:13:42.311 LabApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose.
[I 2024-10-08 13:13:42.314 ServerApp] jupyterlab | extension was successfully linked.
[I 2024-10-08 13:13:42.314 ServerApp] jupyterlab_code_formatter | extension was successfully linked.
[I 2024-10-08 13:13:42.314 ServerApp] jupyterlab_git | extension was successfully linked.
[W 2024-10-08 13:13:42.316 NotebookApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose.
[I 2024-10-08 13:13:42.318 ServerApp] nbclassic | extension was successfully linked.
[I 2024-10-08 13:13:42.318 ServerApp] nbdime | extension was successfully linked.
[W 2024-10-08 13:13:42.319 JupyterNotebookApp] 'extra_template_paths' was found in both NotebookApp and ServerApp. This is likely a recent change. This config will only be set in NotebookApp. Please check if you should also config these traits in ServerApp for your purpose.
[I 2024-10-08 13:13:42.322 ServerApp] notebook | extension was successfully linked.
[I 2024-10-08 13:13:42.322 ServerApp] Writing Jupyter server cookie secret to /tmp/jupyter_cookie_secret
[I 2024-10-08 13:13:42.599 ServerApp] notebook_shim | extension was successfully linked.
[I 2024-10-08 13:13:42.599 ServerApp] panel.io.jupyter_server_extension | extension was successfully linked.
[I 2024-10-08 13:13:42.615 ServerApp] notebook_shim | extension was successfully loaded.
[I 2024-10-08 13:13:42.617 ServerApp] jupyter_lsp | extension was successfully loaded.
[I 2024-10-08 13:13:42.617 ServerApp] jupyter_server_mathjax | extension was successfully loaded.
[I 2024-10-08 13:13:42.633 ServerApp] jupyter_server_proxy | extension was successfully loaded.
[I 2024-10-08 13:13:42.634 ServerApp] jupyter_server_terminals | extension was successfully loaded.
[I 2024-10-08 13:13:42.640 JupyterHubSingleUser] Updating Hub with activity every 300 seconds
[I 2024-10-08 13:13:42.641 ServerApp] jupyterhub | extension was successfully loaded.
[I 2024-10-08 13:13:42.654 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.11/site-packages/jupyterlab
[I 2024-10-08 13:13:42.654 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[I 2024-10-08 13:13:42.656 LabApp] Extension Manager is 'pypi'.
[I 2024-10-08 13:13:42.658 ServerApp] jupyterlab | extension was successfully loaded.
[I 2024-10-08 13:13:42.659 ServerApp] Registered jupyterlab_code_formatter server extension
[I 2024-10-08 13:13:42.659 ServerApp] jupyterlab_code_formatter | extension was successfully loaded.
[I 2024-10-08 13:13:42.664 ServerApp] jupyterlab_git | extension was successfully loaded.
[I 2024-10-08 13:13:42.673 ServerApp] nbclassic | extension was successfully loaded.
[I 2024-10-08 13:13:42.726 ServerApp] nbdime | extension was successfully loaded.
[I 2024-10-08 13:13:42.730 ServerApp] notebook | extension was successfully loaded.
[I 2024-10-08 13:13:42.731 ServerApp] panel.io.jupyter_server_extension | extension was successfully loaded.
[I 2024-10-08 13:13:42.731 ServerApp] Serving notebooks from local directory: /home/jovyan
[I 2024-10-08 13:13:42.731 ServerApp] Jupyter Server 2.7.3 is running at:
[I 2024-10-08 13:13:42.731 ServerApp] http://jupyter-shaun-2eturner1-40nhs-2enet:8888/user/shaun.turner1@nhs.net/lab?token=...
[I 2024-10-08 13:13:42.731 ServerApp]     http://127.0.0.1:8888/user/shaun.turner1@nhs.net/lab?token=...
[I 2024-10-08 13:13:42.731 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 2024-10-08 13:13:43.909 ServerApp] Skipped non-installed server(s): bash-language-server, dockerfile-language-server-nodejs, javascript-typescript-langserver, jedi-language-server, julia-language-server, pyright, python-language-server, python-lsp-server, r-languageserver, sql-language-server, texlab, typescript-language-server, unified-language-server, vscode-css-languageserver-bin, vscode-html-languageserver-bin, vscode-json-languageserver-bin, yaml-language-server
[W 2024-10-08 13:16:12.147 ServerApp] 404 GET / (@127.0.0.1) 12.78ms
[W 2024-10-08 13:18:56.729 ServerApp] 404 GET / (@100.64.0.108) 1.48ms
[W 2024-10-08 13:21:04.087 ServerApp] 404 GET / (@100.64.0.108) 1.35ms
```

**Comment by manics** [comment]
A 404 sounds right- you're  making an unauthenticated request to the `/`, but server is running under `/user/<name>`. The important thing is we know the network isn't a problem, which unfortunately points to a difficult to debug KubeSpawner issue like the one you linked above.

Is there any pattern to which servers have this problem? e.g. large/slow to start images, or only after a certain time period, or number of users/launches?

**Comment by qcaas-nhs-sjt** [comment]
> A 404 sounds right- you're making an unauthenticated request to `/`, but server is running under `/user/<name>`. The important thing is we know the network isn't a problem, which unfortunately points to a difficult to debug KubeSpawner issue like the one you linked above.
> 
> Is there any pattern to which servers have this problem? e.g. large/slow to start images, or only after a certain time period, or number of users/launches?

sadly in each case the image was already place on the machine. There doesn't appear to be a lot of stress on the server either

After mine expired i tried to recreate it and got straight in, same exact settings, etc. I also tried deleting the server a number of times and was presented the same issue

**Comment by qcaas-nhs-sjt** [comment]
@manics have rechecked the jupyterhub logs and found this against mine:

```
[E 2024-10-08 13:29:48.827 JupyterHub reflector:367] Error when watching resources, retrying in 0.2s
    Traceback (most recent call last):
      File "/usr/local/lib/python3.11/site-packages/kubespawner/reflector.py", line 293, in _watch_and_update
        resource_version = await self._list_and_update(resource_version)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/kubespawner/reflector.py", line 236, in _list_and_update
        for p in initial_resources["items"]
                 ~~~~~~~~~~~~~~~~~^^^^^^^^^
    KeyError: 'items'
    
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7fcfe6bbc690>
```

Ironically i didn't see this message against the first one i checked. Does this help at all?

**Comment by manics** [comment]
The only time I've seen that is when someone configures their Z2JH with a custom RBAC role that's missing permissions, presumably resulting in an empty API response from K8s. We could probably add some debug logs to output `initial_resources` though that could be very verbose.

I wonder if `Unclosed client session` could be related- I'd expect all exceptions to be cleaned up, so maybe something is leaking.

**Comment by qcaas-nhs-sjt** [comment]
> The only time I've seen that is when someone configures their Z2JH with a custom RBAC role that's missing permissions, presumably resulting in an empty API response from K8s. We could probably add some debug logs to output `initial_resources` though that could be very verbose.

We don't specify the rbac.create property in jupyter at the moment so it should default to the standard z2jh value, we do add to the permissions, with custom roles that append additional permissions but this appends to jupyters permissions and doesn't take anything away. And if this were a permissions issues then surely permissions should be consistent rather than intermittently responding incorrectly.?
 
> I wonder if `Unclosed client session` could be related- I'd expect all exceptions to be cleaned up, so maybe something is leaking.

This could very well be the case, we have noticed a number of occasions where a container has been left running and not shut down despite the user not having logged into it for a long period of time. I'm unsure of how kubespawner manages sessions though so will need to look into it

**Comment by manics** [comment]
I skimmed through the KubeSpawner codebase:
- The use of singletons that aren't cleaned up is intentional https://github.com/jupyterhub/kubespawner/blob/621be0e68d9dea38ac899c6054f966dd68d1d028/kubespawner/reflector.py#L28-L32 , though we should try and work out whether that's the cause of the `Unclosed client session` above, or whether that's a bug somewhere else
- I found some possible missing `await`s: https://github.com/jupyterhub/kubespawner/issues/858

**Comment by qcaas-nhs-sjt** [comment]
While the Unclosed client session stuff does need addressing I'm seeing it even when we're not seeing the lauch failures above. 

We've restarted all nodes on the prod cluster in the hope that will resolve the issue, we've not seen it yet but at the moment nobody else is using the system so not sure when we can expect to see it again, will keep monitoring for now

**Comment by qcaas-nhs-sjt** [comment]
restart of the nodes did not resolve the issue, we have therefore deleted the pvc for the jupyter hub database and reinstalled jupyter hub under the theory that our database may be corrupted

the system is back up and running, monitoring the logs to see if it occurs again

**Comment by qcaas-nhs-sjt** [comment]
yup still broken, as apparently is the stg environment

**Comment by qcaas-nhs-sjt** [comment]
logs from stg
```
[E 2024-10-10 12:06:05.712 JupyterHub spawner:2810] Pod jupyterhub/jupyter-vishnu-2echandrabalan-40lthtr-2enhs-2euk never showed up in reflector, restarting pod reflector
[E 2024-10-10 12:06:05.712 JupyterHub spawner:2814] Pods: []
[E 2024-10-10 12:06:05.714 JupyterHub gen:630] Exception in Future <Task finished name='Task-17427089' coro=<KubeSpawner._start() done, defined at /usr/local/lib/python3.11/site-packages/kubespawner/spawner.py:2663> exception=TimeoutError('pod jupyterhub/jupyter-vishnu-2echandrabalan-40lthtr-2enhs-2euk did not start in 900 seconds!')> after timeout
    Traceback (most recent call last):
      File "/usr/local/lib/python3.11/site-packages/tornado/gen.py", line 625, in error_callback
        future.result()
      File "/usr/local/lib/python3.11/site-packages/kubespawner/spawner.py", line 2801, in _start
        await exponential_backoff(
      File "/usr/local/lib/python3.11/site-packages/jupyterhub/utils.py", line 237, in exponential_backoff
        raise asyncio.TimeoutError(fail_message)
    TimeoutError: pod jupyterhub/jupyter-vishnu-2echandrabalan-40lthtr-2enhs-2euk did not start in 900 seconds!
```

**Comment by qcaas-nhs-sjt** [comment]
i have tried replicating this issue on stg in multiple different ways but i seem unable to replicate the issue myself

**Comment by manics** [comment]
A couple of ideas: 
- install a second deployment in a different namespace on the same K8S cluster and see if you have the same problem. I presume you will since you're still seeing it after deleting the hub PVC.
- exec into the hub pod and make an unauthenticated curl request to the K8S https API endpoint to verify there's network connectivity between the hub and K8S API
- exec into the hub and make a K8S API request which should use the pod's service account. This is a bit tricky since the `kubernetes_async` python library is installed so any python script you write will need to wrap async call. Alternatively maybe you could copy a `kubectl` binary into the pod and run it?

**Comment by manics** [comment]
Another idea: override `hub.args`
https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/3.3.8/jupyterhub%2Ftemplates%2Fhub%2Fdeployment.yaml#L108-L121
to
`sh -c 'while true; do jupyterhub ... ; done'`
Next time the bug appears exec into the hub pod and kill jupyterhub. The sh loop should automatically restart jupyterhub without restarting the pod. It'll be interesting to know whether this is sufficient, or whether the pod needs to be terminated.

**Comment by vvcb** [comment]
@qcaas-nhs-sjt , @manics - thank you for continuing to look into this. This continues to be an issue and I have spent the last 2 hours trawling k8s logs. The intermittent nature of the problem makes it frustrating - but it is recurrent even though not entirely reproducible. 

The issue appears to that every time I request a workspace from the UI, `hub` actually spins this up almost immediately. I have compared the logs from the `hub`, `chp`, `nginx` (`ngx`)and `single-user` (`nbk`) pods for the failure vs success cases. Below is a screen shot after lining these up. On the left is a failed instance and on the right is a successful instance.

The `nbk` pod never gets registered with `chp`. I am not sure if the `hub` is responsible for this step and if it is, why it has started failing all of a sudden.  I will keep investigating.

![image](https://github.com/user-attachments/assets/8cdeb15a-c6f8-424c-ae89-b95770079771)

On a separate note, there doesn't seem to be any pressure on `etcd`.

![image](https://github.com/user-attachments/assets/2b358f35-ae32-4241-be8c-4ee3a5b638d9)

![image](https://github.com/user-attachments/assets/94b904be-a160-4334-bea7-70723e3b7181)

@qcaas-nhs-sjt , following is the KQL for the logs in that screenshot. 

```kql
ContainerLog
| where TimeGenerated between (datetime('2024-10-10 18:00:00') .. datetime('2024-10-10 21:30:00'))
| where ContainerID  in 
    (
        'b214d581b4fdddd7ef4869f93525e45bb4f2d00052f11f3732072d4f218c87be', 
        'f5478c61d45775d3cdf05439cbb293435b490d76e31cc8d56648a3bc45e32d59', 
        '2e4e6db76999c4e1d0f1e5bdc62539ae4bb870eaad3c587659f303515e86b28c', 
        'c394ce68aa62ab19f5637fbaa9f0d4727c4652b5fb85c62d2bef31dd533546c6',
        '954c8bfe96d613500c26ffe55be1e4b5777f2b06057f0d32aae3618a6a64ca84' 
    )
| union (ContainerLogV2
    | project-rename ContainerID = ContainerId
    | where TimeGenerated between (datetime('2024-10-10 18:00:00') .. datetime('2024-10-10 21:30:00'))
    | where ContainerID  in 
    (
        'b214d581b4fdddd7ef4869f93525e45bb4f2d00052f11f3732072d4f218c87be', 
        'f5478c61d45775d3cdf05439cbb293435b490d76e31cc8d56648a3bc45e32d59', 
        '2e4e6db76999c4e1d0f1e5bdc62539ae4bb870eaad3c587659f303515e86b28c', 
        'c394ce68aa62ab19f5637fbaa9f0d4727c4652b5fb85c62d2bef31dd533546c6',
        '954c8bfe96d613500c26ffe55be1e4b5777f2b06057f0d32aae3618a6a64ca84'
    )    
    )
| order by TimeGenerated asc 
```
Quite a bit of post-processing to line things up. Who said doing this on k8s was a good idea? 😡 😠 😭 🤯

**Comment by vvcb** [comment]
In the logs on the right, the `hub` is clearly registering the `nbk` to `chp` in _L432_ and in the next 2 lines, `chp` acknowledges this.

Presumably. there is a step before this where the `hub` is listening to/querying the control plane for the state of the `nbk` pod it has requested. We don't see this in the logs but this appears to be where things are failing. The hub sometimes discovers the `nbk` pod and sometimes doesn't. Time for bed. @qcaas-nhs-sjt can pull his hair out tomorrow.

**Comment by consideRatio** [comment]
I'm not having a guess on what goes on, but I think making sure its not flakey k8s network policy enforcement could be relevant  by opting out on creating networkpoliicy resources via the jupyterhub helm chart and looking if there are others in the k8s namespace targetting the pods

**Comment by consideRatio** [comment]
Also, is there delay on user server startup from container startup, or is it quick?

**Comment by manics** [comment]
> The nbk pod never gets registered with chp. I am not sure if the hub is responsible for this step and if it is, why it has started failing all of a sudden. I will keep investigating.

The hub is responsible for creating the CHP route, but since the hub doesn't think the server is running it won't do this.


>```
>[E 2024-10-08 13:29:48.827 JupyterHub reflector:367] Error when watching resources, retrying in 0.2s
>    Traceback (most recent call last):
>      File "/usr/local/lib/python3.11/site-packages/kubespawner/reflector.py", line 293, in _watch_and_update
>        resource_version = await self._list_and_update(resource_version)
>                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>      File "/usr/local/lib/python3.11/site-packages/kubespawner/reflector.py", line 236, in _list_and_update
>        for p in initial_resources["items"]
>                 ~~~~~~~~~~~~~~~~~^^^^^^^^^
>```

Does this error regularly occur, or is it a one-off? I'm not completely familiar with the KubeSpawner code, but it watches K8s events to see when relevant resources are created/destroyed. It initially gets a full list (this is where the above error occurs):
https://github.com/jupyterhub/kubespawner/blob/6.2.0/kubespawner/reflector.py#L211-L242
then subsequently watches events for changes. If there's an error it restarts the whole process of requesting all events and setting up a new watch:
https://github.com/jupyterhub/kubespawner/blob/6.2.0/kubespawner/reflector.py#L244-L262

**Comment by qcaas-nhs-sjt** [comment]
have upgraded kubernetes on stg and prod to bring in line with dev to kubernetes 1.29.5 this has not fixed the problem though there is a new error showing

```
[E 2024-10-11 11:02:59.763 JupyterHub reflector:367] Error when watching resources, retrying in 6.4s
    Traceback (most recent call last):
      File "/usr/local/lib/python3.11/site-packages/kubespawner/reflector.py", line 314, in _watch_and_update
        async for watch_event in stream:
      File "/usr/local/lib/python3.11/site-packages/kubernetes_asyncio/watch/watch.py", line 133, in __anext__
        return await self.next()
               ^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/kubernetes_asyncio/watch/watch.py", line 176, in next
        return self.unmarshal_event(line, self.return_type)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.11/site-packages/kubernetes_asyncio/watch/watch.py", line 103, in unmarshal_event
        raise client.exceptions.ApiException(status=obj['code'], reason=reason)
    kubernetes_asyncio.client.exceptions.ApiException: (410)
    Reason: Expired: too old resource version: 201961056 (201961577)
```

I have logged into one of the containers and installed kubectl to query the pods:

```
jovyan@jupyterhub-hub-54795f768f-pj97k:~$ ./kubectl get pods -n jupyterhub 
NAME                                         READY   STATUS    RESTARTS   AGE
jupyter-shaun-2eturner1-40nhs-2enet          1/1     Running   0          7m26s
jupyterhub-continuous-image-puller-5c7f9     1/1     Running   0          47m
jupyterhub-continuous-image-puller-8pfdr     1/1     Running   0          50m
jupyterhub-continuous-image-puller-jm9bk     1/1     Running   0          45m
jupyterhub-hub-54795f768f-pj97k              1/1     Running   0          48m
jupyterhub-proxy-5689cf6ff-wp9gw             1/1     Running   0          48m
jupyterhub-user-scheduler-5f7459695b-2672z   1/1     Running   0          46m
jupyterhub-user-scheduler-5f7459695b-8m4ns   1/1     Running   0          48m
```

secrets:
```
jovyan@jupyterhub-hub-54795f768f-pj97k:~$ ./kubectl get secrets -n jupyterhub 
NAME                               TYPE                 DATA   AGE
jupyter-cookie                     Opaque               1      105d
jupyter-credentials                Opaque               1      105d
jupyter-storage                    Opaque               2      105d
jupyter.xlscsde.nhs.uk-tls         kubernetes.io/tls    2      105d
jupyterhub-hub                     Opaque               5      23h
sh.helm.release.v1.jupyterhub.v1   helm.sh/release.v1   1      23h
```

services:
```
jovyan@jupyterhub-hub-54795f768f-pj97k:~$ ./kubectl get svc -n jupyterhub
NAME                      TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
jupyterhub-hub            ClusterIP   100.64.195.81    <none>        8081/TCP   23h
jupyterhub-proxy-api      ClusterIP   100.64.245.197   <none>        8001/TCP   23h
jupyterhub-proxy-public   ClusterIP   100.64.212.7     <none>        80/TCP     23h
```

I've tried watches on this as well and they all work.

interestingly configmaps are not visible but secrets are:
```
jovyan@jupyterhub-hub-54795f768f-pj97k:~$ ./kubectl get configmap -n jupyterhub 
Error from server (Forbidden): configmaps is forbidden: User "system:serviceaccount:jupyterhub:jupyterhub-hub" cannot list resource "configmaps" in API group "" in the namespace "jupyterhub": Azure does not have opinion for this user.
```

We can also see the analytics workspaces
```
^Cjovyan@jupyterhub-hub-54795f768f-pj97k:~$ ./kubectl get AnalyticsWorkspaces -n jupyterhub 
NAME                                DISPLAYNAME                                            EXPIRES      STATUS        STORAGECLAIM
advanced-generic-workspace          Advanced Generic Workspace                             2030-01-01   Provisioned   jupyter-advanced-generic-workspace
arcnwc-chi-se419                    ARCNWC-CHI-SE419                                       2025-02-22   Provisioned   jupyter-arcnwc-chi-se419
arcnwc-chi-se421                    ARCNWC-CHI-SE421                                       2025-02-22                 
avoidable-admissions                Avoidable Admissions                                   2026-07-01   Provisioned   jupyter-avoidable-admissions
default-generic-workspace           Default Generic Workspace                              2030-01-01   Provisioned   jupyter-default-generic-workspace
diabetic-foot-ulcer-research        Diabetic Foot Ulcer Research                           2030-01-01   Provisioned   jupyter-diabetic-foot-ulcer-research
lsc-maternity-workspace             LSC Maternity Analytics                                2030-12-31   Provisioned   jupyter-lsc-maternity-workspace
lth-data-science-team-workspace     LTH Data Science Team Workspace                        2030-01-01   Provisioned   jupyter-lth-data-science-team-workspace
lth-finance-workspace               LTH Finance Workspace                                  2025-07-31   Provisioned   jupyter-lth-finance-workspace
lth-ugi-surgery-generic-workspace   LTH UGI Surgery Generic Workspace                      2030-01-01   Provisioned   jupyter-lth-ugi-surgery-generic-workspace
neurosurgery-generic-workspace      Neurosurgery Generic Workspace                         2025-12-31                 
nihr-lcrf-workspace                 NIHR Lancashire Clinical Research Facility Workspace   2025-07-31   Provisioned   jupyter-nihr-lcrf-workspace
smoke-test-001                      Smoke Test 001                                         2025-12-31   Provisioned   jupyter-smoke-test-001
uhmb-generic-workspace              UHMB Generic Workspace                                 2030-12-31   Provisioned   jupyter-uhmb-generic-workspace
uom-antimicrobial-workspace         Antimicrobial Prescribing Research                     2030-12-31   Provisioned   jupyter-uom-antimicrobial-workspace
vc-sd-workspace                     VC-SD Workspace                                        2025-12-31   Provisioned   jupyter-vc-sd-workspace
```

and bindings:
```
jovyan@jupyterhub-hub-54795f768f-pj97k:~$ ./kubectl get AnalyticsWorkspaceBindings -n jupyterhub 
NAME                                                    WORKSPACE                           EXPIRES      STATUS   COMMENTS
advanced-generic-workspace-shaun-turner                 advanced-generic-workspace          2030-01-01        
...
```

**Comment by qcaas-nhs-sjt** [comment]
So looking at the code there are a few different reflectors in play:
There are those in the kubespawner.py which are
* [Pods](https://github.com/jupyterhub/kubespawner/blob/452039e5f0da755957ab60758ce9d791ca8d8bb0/kubespawner/spawner.py#L57)
* [Events](https://github.com/jupyterhub/kubespawner/blob/452039e5f0da755957ab60758ce9d791ca8d8bb0/kubespawner/spawner.py#L79)

and those in proxy:
* [Ingresses](https://github.com/jupyterhub/kubespawner/blob/452039e5f0da755957ab60758ce9d791ca8d8bb0/kubespawner/proxy.py#L18)
* [Services](https://github.com/jupyterhub/kubespawner/blob/452039e5f0da755957ab60758ce9d791ca8d8bb0/kubespawner/proxy.py#L27)
* [Endpoints](https://github.com/jupyterhub/kubespawner/blob/452039e5f0da755957ab60758ce9d791ca8d8bb0/kubespawner/proxy.py#L35)

As the logs don't actually seem to indicate which reflector is at play i'm going to assume it needs all of these reflectors to be working

the containers cannot query endpoints or ingresses:
```
jovyan@jupyterhub-hub-54795f768f-pj97k:~$ ./kubectl get ingress
Error from server (Forbidden): ingresses.networking.k8s.io is forbidden: User "system:serviceaccount:jupyterhub:jupyterhub-hub" cannot list resource "ingresses" in API group "networking.k8s.io" in the namespace "jupyterhub": Azure does not have opinion for this user.
jovyan@jupyterhub-hub-54795f768f-pj97k:~$ ./kubectl get endpoints
Error from server (Forbidden): endpoints is forbidden: User "system:serviceaccount:jupyterhub:jupyterhub-hub" cannot list resource "endpoints" in API group "" in the namespace "jupyterhub": Azure does not have opinion for this user.
```

To address this i'm going to give the service account temporary permissions to get/watch these items

**Comment by manics** [comment]
Can you patch reflector.py to add some debug logging, such as
https://github.com/manics/kubespawner/commit/df3a6e53cb8ecad96903746d7aa0368e68cbdef0

This will be quite verbose, but should include
```py
hub-95588d8fd-hgbq5 hub [D 2024-10-11 13:58:33.221 JupyterHub reflector:306] Connecting pods watcher
hub-95588d8fd-hgbq5 hub [D 2024-10-11 13:58:33.225 JupyterHub reflector:232] _list_and_update pods list_namespaced_pod {'label_selector': 'component=singleuser-server', 'field_selector': '', '_request_timeout': 60, '_preload_content': False, 'resource_version': '115996435', 'resource_version_match': 'NotOlderThan', 'namespace': 'jupyter'}
hub-95588d8fd-hgbq5 hub [D 2024-10-11 13:58:33.234 JupyterHub reflector:250] _list_and_update initial_resources: {'kind': 'PodList', 'apiVersion': 'v1', 'metadata': {'resourceVersion': '115996489'}, 'items': ...
```
and
```py
hub-95588d8fd-hgbq5 hub [D 2024-10-11 13:58:33.234 JupyterHub reflector:330] _watch_and_update watch pods list_namespaced_pod {'label_selector': 'component=singleuser-server', 'field_selector': '', 'resource_version': '115996489', 'namespace': 'jupyter', '_request_timeout': 60, 'timeout_seconds': 10}
hub-95588d8fd-hgbq5 hub [D 2024-10-11 13:58:43.238 JupyterHub reflector:393] pods watcher timeout
```

I'd like to trace though the `resourceVersion`- check that they're not being corrupted somewhere along the line.

**Comment by qcaas-nhs-sjt** [comment]
We've enabled the AKS Audit tables in azure diagnostics for the production environment which has allowed us to see the following queries from jupyter hub, as you can see the 504 errors are fairly common. Interestingly when a 504 error occurs it doesn't stop trying to poll and reset but instead keeps polling for the same resource version which it has already been told is too large.

[query_data(1).csv](https://github.com/user-attachments/files/17361964/query_data.1.csv)

I can see from the logs that this is also happening for other services, but those services such as keda, these services seem to be recovering themselves by handling the exceptions and some retry logic.

**Comment by manics** [comment]
Debug logging needs to be configured at the application level, either using [Z2JH config](https://z2jh.jupyter.org/en/stable/administrator/debug.html#verbose-logging)
```yaml
debug:
  enabled: true
```
or Traitlets:
```yaml
hub:
  config:
    Application:
      log_level: DEBUG
```

**Comment by manics** [comment]
KubeSpawner _should_ already handle the invalid `resource_version` error:
`_list_and_update` throws a `client.ApiException`:
https://github.com/jupyterhub/kubespawner/blob/452039e5f0da755957ab60758ce9d791ca8d8bb0/kubespawner/reflector.py#L232-L245

`client.ApiException` is a subclass for `Exception`:
```py
>>> from kubernetes_asyncio import client
>>> client.ApiException.mro()
[<class 'kubernetes_asyncio.client.exceptions.ApiException'>, <class 'kubernetes_asyncio.client.exceptions.OpenApiException'>, <class 'Exception'>, <class 'BaseException'>, <class 'object'>]
```

`_watch_and_update` calls `_list_and_update` inside a loop:
https://github.com/jupyterhub/kubespawner/blob/452039e5f0da755957ab60758ce9d791ca8d8bb0/kubespawner/reflector.py#L308
It should catch the `client.ApiException`, and reset `resouce_version = "0"`
https://github.com/jupyterhub/kubespawner/blob/452039e5f0da755957ab60758ce9d791ca8d8bb0/kubespawner/reflector.py#L372-L386

Obviously something has gone wrong.

**Comment by manics** [comment]
Someone's figured it out- just a simple logical error in kubespawner!
https://github.com/jupyterhub/kubespawner/issues/868

**Comment by vvcb** [comment]
@qcaas-nhs-sjt , I have not seen this issue again in the past 2 months and no complaints from the students either. Bizarre. Should we close this and reopen if this crops up again?
